tnet/listener.go

```diff
@@ -0,0 +1,95 @@
+// Copyright (c) 2015 Uber Technologies, Inc.
+
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package tnet
+
+import (
+	"net"
+	"sync"
+)
+
+// WrapListener returns a new Listener around the provided net.Listener.
+// The returned Listener has a guarantee that when Close returns, it will no longer
+// accept any new connections.
+// See: https://github.com/uber/tchannel-go/issues/141
+func WrapListener(l net.Listener) net.Listener {
+	return &listener{Listener: l, cond: sync.NewCond(&sync.Mutex{})}
+}
```

Comment:
(Moved to main PR instead of commit - sorry, getting used to GitHub again.)

Nit: could this function be called `Wrap`? The type system already tells callers that it's wrapping a `net.Listener`.